chain-impl-mockchain/src/testing/arbitrary/transaction.rs

```diff
@@ -0,0 +1,197 @@
+use crate::ledger::Ledger;
+use crate::testing::arbitrary::transaction::rand::prelude::SliceRandom;
+use crate::transaction::{Input, Output};
+use crate::utxo::{self, Entry};
+use crate::{
+    testing::address::{AddressData, AddressDataValue},
+    value::*,
+};
+use chain_addr::Address;
+use quickcheck::{Arbitrary, Gen};
+use std::iter;
+
+extern crate rand;
+use super::ArbitraryAddressDataVec;
+use rand::Rng;
+
+#[derive(Clone, Debug)]
+pub struct AribtraryValidTransactionData {
+    pub addresses: Vec<AddressDataValue>,
+    input_addresses: Vec<AddressDataValue>,
+    output_addresses: Vec<AddressDataValue>,
+}
+
+impl Arbitrary for AribtraryValidTransactionData {
+    fn arbitrary<G: Gen>(gen: &mut G) -> Self {
+        AribtraryValidTransactionDataBuilder::new(ArbitraryAddressDataVec::arbitrary(gen).0).build()
+    }
+}
+
+pub struct AribtraryValidTransactionDataBuilder {
+    source: Vec<AddressData>,
+}
+
+impl AribtraryValidTransactionDataBuilder {
+    pub fn new(source: Vec<AddressData>) -> Self {
+        AribtraryValidTransactionDataBuilder { source: source }
+    }
+
+    pub fn build(&self) -> AribtraryValidTransactionData {
+        let values: Vec<Value> = Self::generate_random_values()
+            .take(self.source.len())
+            .collect();
+
+        let addresses_values: Vec<AddressDataValue> =
+            Self::zip_addresses_and_values(&self.source, values);
```

Comment:
So the list of addresses is always the same and in the same order, only the values are random between invocations of `build` on the same builder instance? Is this the intent here?

Reply:
Hm, no. list of addresses is also arbitraty because its generated from ArbitraryValidTransactionData

Reply:
Yes, but it will be always the same from this instance of `...Builder`, because it is generated once and then baked in? If you don't use the builder to generate transaction data repeatedly, why even have the separate builder type?

Reply:
ok. removed the builder.